lib/vm/src/instance.rs

```diff
@@ -761,110 +661,272 @@ impl Instance {
+        self.instance == other.instance
+    }
+}
+
+impl Drop for Instance {
+    fn drop(&mut self) {
+        println!("Dropping `wasmer_vm::Instance`");
+    }
+}
+
+impl Drop for InstanceAllocator {
+    /// Drop the `InstanceAllocator`.
     ///
-    /// # Safety
-    /// - `instance_ptr` must point to enough memory that all of the offsets in
-    ///   `offsets` point to valid locations in memory.
-    pub unsafe fn table_definition_locations(
-        instance_ptr: NonNull<u8>,
-        offsets: &VMOffsets,
-    ) -> Vec<NonNull<VMTableDefinition>> {
-        let num_tables = offsets.num_local_tables;
-        let num_tables = usize::try_from(num_tables).unwrap();
-        let mut out = Vec::with_capacity(num_tables);
-        // TODO: better encapsulate this logic, this shouldn't be duplicated
-        let base_ptr = instance_ptr.as_ptr().add(std::mem::size_of::<Instance>());
-        for i in 0..num_tables {
-            let table_offset = offsets.vmctx_vmtable_definition(LocalTableIndex::new(i));
-            let table_offset = usize::try_from(table_offset).unwrap();
+    /// This will decrement the strong reference count. If it reaches
+    /// 1, then the `Self.instance` will be deallocated with
+    /// `Self::deallocate_instance`.
+    fn drop(&mut self) {
+        println!("Trying to drop `wasmer_vm::InstanceAllocator`...");
 
-            let new_ptr = NonNull::new_unchecked(base_ptr.add(table_offset));
+        // Because `fetch_sub` is already atomic, we do not need to
+        // synchronize with other threads unless we are going to
+        // delete the object.
+        if self.strong.fetch_sub(1, atomic::Ordering::Release) != 1 {
+            println!("... not now");
 
-            out.push(new_ptr.cast());
+            return;
         }
-        out
+
+        println!("Yep, dropping it.");
+
+        // This fence is needed to prevent reordering of use of the data and
+        // deletion of the data. Because it is marked `Release`, the decreasing
+        // of the reference count synchronizes with this `Acquire` fence. This
+        // means that use of the data happens before decreasing the reference
+        // count, which happens before this fence, which happens before the
+        // deletion of the data.
+        //
+        // As explained in the [Boost documentation][1],
+        //
+        // > It is important to enforce any possible access to the object in one
+        // > thread (through an existing reference) to *happen before* deleting
+        // > the object in a different thread. This is achieved by a "release"
+        // > operation after dropping a reference (any access to the object
+        // > through this reference must obviously happened before), and an
+        // > "acquire" operation before deleting the object.
+        //
+        // [1]: https://www.boost.org/doc/libs/1_55_0/doc/html/atomic/usage_examples.html
+        atomic::fence(atomic::Ordering::Acquire);
```

Comment:
When we do this, don't we have to check that it wasn't incremented? 🤔

Reply:
This code is from `Arc` itself. I trust them more than I trust myself.

Reply:
One thing to keep in mind is that this code is not doing the same thing as `Arc`, I'm unclear on if it's different enough to matter though.

Notably the std::arc::Arc source code defers actually deallocating the underlying memory until all the `Weak` pointers are dropped.

And that code does similar checks:

```Rust
        if inner.weak.fetch_sub(1, Release) == 1 {
            acquire!(inner.weak);
            unsafe { Global.dealloc(self.ptr.cast(), Layout::for_value(self.ptr.as_ref())) }
        }
```

After reviewing `InstanceAllocator` again, I think it's probably similar enough that we can call it correct!

Reply:
:+1: